Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/qrcode/coil/BitMatrixDecoder.kt

```diff
@@ -0,0 +1,100 @@
+package de.rki.coronawarnapp.util.qrcode.coil
+
+import android.content.Context
+import android.content.res.Resources
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.drawable.BitmapDrawable
+import coil.bitmap.BitmapPool
+import coil.decode.DecodeResult
+import coil.decode.Decoder
+import coil.decode.Options
+import coil.size.OriginalSize
+import coil.size.PixelSize
+import coil.size.Size
+import dagger.Reusable
+import de.rki.coronawarnapp.util.di.AppContext
+import okio.BufferedSource
+import timber.log.Timber
+import java.io.ObjectInputStream
+import javax.inject.Inject
+import kotlin.math.min
+import kotlin.math.nextUp
+
+@Reusable
+class BitMatrixDecoder @Inject constructor(
+    @AppContext private val context: Context,
+) : Decoder {
+    private val resources: Resources = context.resources
+
+    override fun handles(source: BufferedSource, mimeType: String?): Boolean = mimeType == MIME_TYPE
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    override suspend fun decode(pool: BitmapPool, source: BufferedSource, size: Size, options: Options): DecodeResult {
+        val bitmapStart = System.currentTimeMillis()
+
+        val matrix = ObjectInputStream(source.inputStream()).use {
+            it.readObject() as Array<ByteArray>
+        }
+        val (matrixHeight, matrixWidth) = run {
+            //  The matrix is array[y][x], e.g. a list of columns
+            matrix.size to matrix.first().size
+        }
+
+        val (targetWidth, targetHeight) = when (size) {
+            OriginalSize -> matrixWidth to matrixWidth
+            is PixelSize -> size.width to size.height
+        }
+
+        val marginFactor = 1
+        val squareSize = min(
+            targetWidth.toFloat() / (matrixWidth + marginFactor * 2),
+            targetHeight.toFloat() / (matrixHeight + marginFactor * 2)
+        ).nextUp()
```

Comment:
what the margin here for ?
for specific ⬛  or for whole QrCode?
I think the margin can be handled by axing already

Reply:
We no longer use zxing for generating the visual representation. We only use it to get the matrix data.
The margin is the size of a qrcode square.

So a margin factor of 1 means that there is at least white space equal to the size of 1 qrcode "pixel" around the whole qrcode.

I played around with it a bit, and if the qrcode is on dark background, we need a small border around the qrcode for it to be scanned reliably. Likely because the dark background touching the black orientation squares in the corners of the qrcode.

So the produced image has the minimum margin for it to be able to be scanned. Any extra margin should be added via `android:margin/padding` in the UI.